File: companies/HOJAI-AI/sdk/flutter/lib/genie_client.dart
import 'dart:async';
import 'dart:convert';
import 'dart:io';
import 'package:http/http.dart' as http;

/// Error types for the Genie Flutter SDK
class GenieException implements Exception {
  final String message;
  final String code;
  GenieException(this.message, [this.code = 'UNKNOWN']);

  @override
  String toString() => 'GenieException($code): $message';
}

class GenieNetworkException extends GenieException {
  GenieNetworkException(super.message) : super('NETWORK');
}

class GenieAuthException extends GenieException {
  GenieAuthException(super.message) : super('AUTH');
}

class GenieApiException extends GenieException {
  GenieApiException(super.message, [super.code = 'API']);
}

/// Response models

class GenieResponse {
  final String responseId;
  final String text;
  final List<GenieAction>? actions;
  final Map<String, dynamic>? context;
  final DateTime? timestamp;

  GenieResponse({
    required this.responseId,
    required this.text,
    this.actions,
    this.context,
    this.timestamp,
  });

  factory GenieResponse.fromJson(Map<String, dynamic> json) {
    return GenieResponse(
      responseId: json['responseId'] ?? '',
      text: json['text'] ?? '',
      actions: (json['actions'] as List?)?.map((a) => GenieAction.fromJson(a)).toList(),
      context: json['context'],
      timestamp: json['timestamp'] != null ? DateTime.tryParse(json['timestamp']) : null,
    );
  }
}

class GenieAction {
  final String type;
  final Map<String, dynamic>? payload;

  GenieAction({required this.type, this.payload});

  factory GenieAction.fromJson(Map<String, dynamic> json) {
    return GenieAction(type: json['type'] ?? '', payload: json['payload']);
  }
}

enum BriefingType { morning, evening, weekly }

class Briefing {
  final BriefingType type;
  final String title;
  final List<BriefingSection> sections;
  final DateTime generatedAt;

  Briefing({
    required this.type,
    required this.title,
    required this.sections,
    required this.generatedAt,
  });

  factory Briefing.fromJson(Map<String, dynamic> json) {
    return Briefing(
      type: BriefingType.values.firstWhere(
        (t) => t.name == (json['type'] ?? 'morning'),
        orElse: () => BriefingType.morning,
      ),
      title: json['title'] ?? '',
      sections: (json['sections'] as List?)
              ?.map((s) => BriefingSection.fromJson(s))
              .toList() ??
          [],
      generatedAt: DateTime.tryParse(json['generatedAt'] ?? '') ?? DateTime.now(),
    );
  }
}

class BriefingSection {
  final String title;
  final String content;
  final String? icon;

  BriefingSection({required this.title, required this.content, this.icon});

  factory BriefingSection.fromJson(Map<String, dynamic> json) {
    return BriefingSection(
      title: json['title'] ?? '',
      content: json['content'] ?? '',
      icon: json['icon'],
    );
  }
}

class Memory {
  final String id;
  final String content;
  final String type;
  final DateTime createdAt;
  final List<String>? tags;
  final double? relevance;

  Memory({
    required this.id,
    required this.content,
    required this.type,
    required this.createdAt,
    this.tags,
    this.relevance,
  });

  factory Memory.fromJson(Map<String, dynamic> json) {
    return Memory(
      id: json['id'] ?? '',
      content: json['content'] ?? '',
      type: json['type'] ?? 'note',
      createdAt: DateTime.tryParse(json['createdAt'] ?? '') ?? DateTime.now(),
      tags: (json['tags'] as List?)?.cast<String>(),
      relevance: (json['relevance'] as num?)?.toDouble(),
    );
  }
}

class CalendarEvent {
  final String id;
  final String title;
  final DateTime startTime;
  final DateTime endTime;
  final String? location;
  final List<String>? attendees;
  final String? description;

  CalendarEvent({
    required this.id,
    required this.title,
    required this.startTime,
    required this.endTime,
    this.location,
    this.attendees,
    this.description,
  });

  factory CalendarEvent.fromJson(Map<String, dynamic> json) {
    return CalendarEvent(
      id: json['id'] ?? '',
      title: json['title'] ?? '',
      startTime: DateTime.tryParse(json['startTime'] ?? '') ?? DateTime.now(),
      endTime: DateTime.tryParse(json['endTime'] ?? '') ?? DateTime.now(),
      location: json['location'],
      attendees: (json['attendees'] as List?)?.cast<String>(),
      description: json['description'],
    );
  }
}

/// Voice configuration
class VoiceConfig {
  final int sampleRate;
  final String codec;
  final String language;
  final bool wakeWordEnabled;

  VoiceConfig({
    this.sampleRate = 16000,
    this.codec = 'pcm_s16le',
    this.language = 'en-US',
    this.wakeWordEnabled = true,
  });

  Map<String, dynamic> toJson() => {
        'sampleRate': sampleRate,
        'codec': codec,
        'language': language,
        'wakeWordEnabled': wakeWordEnabled,
      };
}

/// Voice session controller for Flutter
class VoiceSessionController {
  final GenieClient _client;
  final VoiceConfig config;
  final StreamController<String> _transcriptController = StreamController<String>.broadcast();
  final StreamController<GenieResponse> _responseController = StreamController<GenieResponse>.broadcast();
  final StreamController<String> _errorController = StreamController<String>.broadcast();
  bool _isActive = false;

  VoiceSessionController(this._client, this.config);

  Stream<String> get transcriptStream => _transcriptController.stream;
  Stream<GenieResponse> get responseStream => _responseController.stream;
  Stream<String> get errorStream => _errorController.stream;
  bool get isActive => _isActive;

  void _emitTranscript(String text) => _transcriptController.add(text);
  void _emitResponse(GenieResponse response) => _responseController.add(response);
  void _emitError(String error) => _errorController.add(error);

  void dispose() {
    _transcriptController.close();
    _responseController.close();
    _errorController.close();
  }
}

/// Main Genie Flutter Client
class GenieClient {
  static final GenieClient _instance = GenieClient._();
  factory GenieClient() => _instance;
  GenieClient._();

  String _baseUrl = 'https://api.hojai.ai';
  String? _apiKey;
  String? _authToken;
  final http.Client _httpClient = http.Client();

  VoiceSessionController? _activeVoiceSession;

  // ── Configuration ──────────────────────────────────────────────────────

  Future<void> configure({required String apiKey, String? baseUrl}) async {
    _apiKey = apiKey;
    if (baseUrl != null) _baseUrl = baseUrl.trimEnd('/');
  }

  void setAuthToken(String token) => _authToken = token;

  String get baseUrl => _baseUrl;

  // ── Core API ────────────────────────────────────────────────────────────

  Future<GenieResponse> sendMessage(String text, {Map<String, dynamic>? context}) async {
    final uri = Uri.parse('$_baseUrl/api/genie/chat');
    final body = {'question': text, if (context != null) 'context': context};

    final response = await _post(uri, body);
    return GenieResponse.fromJson(_decode(response));
  }

  Future<Briefing> getBriefing(BriefingType type) async {
    final uri = Uri.parse('$_baseUrl/api/genie/briefing?type=${type.name}');
    final response = await _get(uri);
    final data = _decode(response);
    // Backend wraps in { data: Briefing }
    if (data is Map && data.containsKey('data')) {
      return Briefing.fromJson(data['data']);
    }
    return Briefing.fromJson(data);
  }

  Future<List<Memory>> searchMemories(String query, {int limit = 10}) async {
    final uri = Uri.parse('$_baseUrl/api/genie/memory/search?q=${Uri.encodeComponent(query)}&limit=$limit');
    final response = await _get(uri);
    final data = _decode(response);
    final memories = data is Map && data.containsKey('memories')
        ? (data['memories'] as List)
        : (data is List ? data : []);
    return memories.map((m) => Memory.fromJson(m)).toList();
  }

  Future<List<CalendarEvent>> getCalendarEvents(DateTime from, DateTime to) async {
    final formatter = 'yyyy-MM-ddTHH:mm:ssZ';
    final fromStr = from.toUtc().toIso8601String();
    final toStr = to.toUtc().toIso8601String();
    final uri = Uri.parse('$_baseUrl/api/genie/calendar/events?from=$fromStr&to=$toStr');
    final response = await _get(uri);
    final data = _decode(response);
    final events = data is Map && data.containsKey('events')
        ? (data['events'] as List)
        : (data is List ? data : []);
    return events.map((e) => CalendarEvent.fromJson(e)).toList();
  }

  Future<Map<String, dynamic>> getUser() async {
    final uri = Uri.parse('$_baseUrl/api/genie/user');
    final response = await _get(uri);
    final data = _decode(response);
    return data is Map ? data : {};
  }

  Future<void> updatePreferences(Map<String, dynamic> prefs) async {
    final uri = Uri.parse('$_baseUrl/api/genie/user/preferences');
    await _put(uri, prefs);
  }

  // ── Voice Session ────────────────────────────────────────────────────────

  Future<VoiceSessionController> startVoiceSession(VoiceConfig config) async {
    final uri = Uri.parse('$_baseUrl/api/genie/voice/session/start');
    await _post(uri, config.toJson());

    final controller = VoiceSessionController(this, config);
    _activeVoiceSession = controller;
    return controller;
  }

  Future<void> stopVoiceSession() async {
    if (_activeVoiceSession == null) return;
    final uri = Uri.parse('$_baseUrl/api/genie/voice/session/end');
    try {
      await _post(uri, {});
    } finally {
      _activeVoiceSession?.dispose();
      _activeVoiceSession = null;
    }
  }

  // ── HTTP Helpers ────────────────────────────────────────────────────────

  Future<http.Response> _get(Uri uri) => _request(uri, 'GET');
  Future<http.Response> _post(Uri uri, Map<String, dynamic> body) =>
      _request(uri, 'POST', body: body);
  Future<http.Response> _put(Uri uri, Map<String, dynamic> body) =>
      _request(uri, 'PUT', body: body);

  Future<http.Response> _request(
    Uri uri,
    String method, {
    Map<String, dynamic>? body,
  }) async {
    final token = _authToken ?? _apiKey ?? '';
    final headers = {
      'Content-Type': 'application/json',
      'Accept': 'application/json',
      'Authorization': 'Bearer $token',
    };

    http.Response response;
    try {
      switch (method) {
        case 'GET':
          response = await _httpClient.get(uri, headers: headers).timeout(const Duration(seconds: 30));
          break;
        case 'POST':
          response = await _httpClient.post(uri, headers: headers, body: jsonEncode(body)).timeout(const Duration(seconds: 30));
          break;
        case 'PUT':
          response = await _httpClient.put(uri, headers: headers, body: jsonEncode(body)).timeout(const Duration(seconds: 30));
          break;
        default:
          response = await _httpClient.get(uri, headers: headers).timeout(const Duration(seconds: 30));
      }
    } on http.ClientException catch (e) {
      throw GenieNetworkException('Network error: ${e.message}');
    }

    if (response.statusCode == 401) throw GenieAuthException('Invalid or expired token');
    if (response.statusCode == 403) throw GenieAuthException('Insufficient permissions');
    if (response.statusCode == 429) throw GenieApiException('Rate limit exceeded', 'RATE_LIMIT');
    if (response.statusCode >= 500) throw GenieApiException('Server error: ${response.statusCode}', 'SERVER_ERROR');
    if (response.statusCode < 200 || response.statusCode >= 300) {
      throw GenieApiException('Request failed: ${response.statusCode}', 'HTTP_${response.statusCode}');
    }

    return response;
  }

  Map<String, dynamic> _decode(http.Response response) {
    if (response.body.isEmpty) return {};
    try {
      final decoded = jsonDecode(response.body);
      if (decoded is Map<String, dynamic>) return decoded;
      if (decoded is List) return {'list': decoded};
      return {'data': decoded};
    } catch (_) {
      return {};
    }
  }

  void dispose() {
    _httpClient.close();
    _activeVoiceSession?.dispose();
  }
}

File: companies/HOJAI-AI/sdk/android/src/main/java/com/hojai/genie/GenieClient.kt
package com.hojai.genie

import kotlinx.coroutines.*
import kotlinx.serialization.Serializable
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.*
import okhttp3.*
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.RequestBody.Companion.toRequestBody
import java.io.IOException
import java.time.Instant
import java.util.UUID
import java.util.concurrent.TimeUnit

// ============================================================================
// Error Types
// ============================================================================

sealed class GenieError(message: String, val code: String = "UNKNOWN") : Exception(message) {
    class NetworkError(underlying: IOException) : GenieError(underlying.message ?: "Network error", "NETWORK")
    class AuthError(message: String) : GenieError(message, "AUTH")
    class ApiError(code: String, message: String) : GenieError(message, code)
    class EncodingError(message: String) : GenieError(message, "ENCODING")
    class DecodingError(message: String, val underlying: Throwable) : GenieError(message, "DECODING")
    class TimeoutError : GenieError("Request timed out", "TIMEOUT")
}

// ============================================================================
// Data Models
// ============================================================================

@Serializable
data class GenieResponse(
    val responseId: String,
    val text: String,
    val actions: List<GenieAction>? = null,
    val context: Map<String, JsonElement>? = null,
    val timestamp: String? = null
)

@Serializable
data class GenieAction(
    val type: String,
    val payload: Map<String, JsonElement>? = null
)

@Serializable
enum class BriefingType { morning, evening, weekly }

@Serializable
data class Briefing(
    val type: BriefingType,
    val title: String,
    val sections: List<BriefingSection>,
    val generatedAt: String
)

@Serializable
data class BriefingSection(
    val title: String,
    val content: String,
    val icon: String? = null
)

@Serializable
data class Memory(
    val id: String,
    val content: String,
    val type: String,
    val createdAt: String,
    val tags: List<String>? = null,
    val relevance: Double? = null
)

@Serializable
data class CalendarEvent(
    val id: String,
    val title: String,
    val startTime: String,
    val endTime: String,
    val location: String? = null,
    val attendees: List<String>? = null,
    val description: String? = null
) {
    fun startMillis(): Long = try { Instant.parse(startTime).toEpochMilli() } catch (_: Exception) { 0L }
    fun endMillis(): Long = try { Instant.parse(endTime).toEpochMilli() } catch (_: Exception) { 0L }
}

@Serializable
data class UserProfile(
    val id: String,
    val email: String,
    val name: String,
    val onboardingComplete: Boolean,
    val preferences: Map<String, JsonElement>? = null,
    val goals: List<String>? = null,
    val createdAt: String
)

// ============================================================================
// Voice Configuration
// ============================================================================

data class VoiceConfig(
    val sampleRate: Int = 16000,
    val codec: String = "pcm_s16le",
    val language: String = "en-US",
    val wakeWordEnabled: Boolean = true
)

// ============================================================================
// Voice Listener
// ============================================================================

interface VoiceListener {
    fun onTranscript(text: String)
    fun onAudioData(data: ByteArray)
    fun onResponse(response: GenieResponse)
    fun onError(error: GenieError)
}

// ============================================================================
// Genie Client
// ============================================================================

class GenieClient private constructor() {

    companion object {
        val instance: GenieClient by lazy { GenieClient() }
    }

    private val json = Json { ignoreUnknownKeys = true; isLenient = true; encodeDefaults = true }
    private val mediaTypeJson = "application/json; charset=utf-8".toMediaType()

    private var baseUrl: String = "https://api.hojai.ai"
    private var apiKey: String = ""
    private var authToken: String? = null

    private val httpClient = OkHttpClient.Builder()
        .connectTimeout(30, TimeUnit.SECONDS)
        .readTimeout(60, TimeUnit.SECONDS)
        .writeTimeout(60, TimeUnit.SECONDS)
        .build()

    private val scope = CoroutineScope(Dispatchers.IO + SupervisorJob())

    // Voice state
    private var isVoiceSessionActive = false
    private var currentVoiceListener: VoiceListener? = null

    // ── Configuration ──────────────────────────────────────────────────────

    fun configure(apiKey: String, baseUrl: String = "https://api.hojai.ai") {
        this.apiKey = apiKey
        this.baseUrl = baseUrl.trimEnd('/')
    }

    fun setAuthToken(token: String) {
        this.authToken = token
    }

    private fun buildRequest(path: String, method: String, body: String? = null): Request {
        val url = "$baseUrl$path"
        val builder = Request.Builder().url(url)
            .addHeader("Accept", "application/json")
            .addHeader("Content-Type", "application/json")

        val token = authToken ?: apiKey
        builder.addHeader("Authorization", "Bearer $token")

        when (method.uppercase()) {
            "GET" -> builder.get()
            "POST" -> builder.post((body ?: "{}").toRequestBody(mediaTypeJson))
            "PUT" -> builder.put((body ?: "{}").toRequestBody(mediaTypeJson))
            "PATCH" -> builder.patch((body ?: "{}").toRequestBody(mediaTypeJson))
            "DELETE" -> builder.delete()
            else -> builder.get()
        }
        return builder.build()
    }

    private suspend fun <T> executeRequest(request: Request, clazz: Class<T>): T = suspendCatching {
        val response = httpClient.newCall(request).await()
        val body = response.body?.string() ?: ""

        when (response.code) {
            in 200..299 -> {
                try {
                    json.decodeFromString(clazz.kotlin.javaObjectType, body)
                } catch (e: Exception) {
                    throw GenieError.DecodingError("Failed to decode response: ${e.message}", e)
                }
            }
            401 -> throw GenieError.AuthError("Invalid or expired token")
            403 -> throw GenieError.AuthError("Insufficient permissions")
            429 -> throw GenieError.ApiError("RATE_LIMIT", "Rate limit exceeded")
            in 500..599 -> throw GenieError.ApiError("SERVER_ERROR", "Internal server error")
            else -> throw GenieError.ApiError("HTTP_${response.code}", "Unexpected response: ${response.code}")
        }
    }.getOrThrow()

    // ── Chat ─────────────────────────────────────────────────────────────────

    suspend fun sendMessage(text: String, context: Map<String, Any>? = null): GenieResponse = scope.e {
        val bodyMap = mutableMapOf("question" to JsonPrimitive(text))
        if (context != null) {
            bodyMap["context"] = json.encodeToJsonElement(context)
        }
        val body = json.encodeToString(JsonObject(bodyMap))
        val request = buildRequest("/api/genie/chat", "POST", body)
        executeRequest(request, GenieResponse::class.java)
    }

    // ── Briefing ─────────────────────────────────────────────────────────────

    suspend fun getBriefing(type: BriefingType): Briefing = scope.e {
        val request = buildRequest("/api/genie/briefing?type=${type.name.lowercase()}", "GET")
        executeRequest(request, Briefing::class.java)
    }

    // ── Memory ───────────────────────────────────────────────────────────────

    suspend fun searchMemories(query: String, limit: Int = 10): List<Memory> = scope.e {
        val url = "/api/genie/memory/search?q=${java.net.URLEncoder.encode(query, "UTF-8")}&limit=$limit"
        val request = buildRequest(url, "GET")
        val wrapper = executeRequest(request, MemoriesWrapper::class.java)
        wrapper.memories
    }

    @Serializable
    private data class MemoriesWrapper(val memories: List<Memory>)

    // ── Calendar ────────────────────────────────────────────────────────────

    suspend fun getCalendarEvents(fromMillis: Long, toMillis: Long): List<CalendarEvent> = scope.e {
        val formatter = java.text.SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss'Z'", java.util.Locale.US).apply {
            timeZone = java.util.TimeZone.getTimeZone("UTC")
        }
        val from = formatter.format(java.util.Date(fromMillis))
        val to = formatter.format(java.util.Date(toMillis))
        val url = "/api/genie/calendar/events?from=$from&to=$to"
        val request = buildRequest(url, "GET")
        val wrapper = executeRequest(request, EventsWrapper::class.java)
        wrapper.events
    }

    @Serializable
    private data class EventsWrapper(val events: List<CalendarEvent>)

    // ── Voice Session ────────────────────────────────────────────────────────

    suspend fun startVoiceSession(config: VoiceConfig, listener: VoiceListener) = scope.e {
        currentVoiceListener = listener
        isVoiceSessionActive = true

        // Notify backend
        val body = json.encodeToString(JsonObject(mapOf(
            "sampleRate" to JsonPrimitive(config.sampleRate),
            "codec" to JsonPrimitive(config.codec),
            "language" to JsonPrimitive(config.language)
        )))
        val request = buildRequest("/api/genie/voice/session/start", "POST", body)
        executeRequest<Unit>(request, Unit::class.java)
    }

    suspend fun stopVoiceSession() = scope.e {
        if (!isVoiceSessionActive) return@e
        isVoiceSessionActive = false
        currentVoiceListener = null

        val request = buildRequest("/api/genie/voice/session/end", "POST")
        try {
            executeRequest<Unit>(request, Unit::class.java)
        } catch (_: Exception) {
            // best effort
        }
    }

    // ── User ─────────────────────────────────────────────────────────────────

    suspend fun getUserProfile(): UserProfile = scope.e {
        val request = buildRequest("/api/genie/user", "GET")
        executeRequest(request, UserProfile::class.java)
    }

    suspend fun updatePreferences(prefs: Map<String, Any>) = scope.e {
        val body = json.encodeToString(JsonObject(prefs.mapValues { json.encodeToJsonElement(it.value) }))
        val request = buildRequest("/api/genie/user/preferences", "PUT", body)
        executeRequest<Unit>(request, Unit::class.java)
    }

    // ── Convenience ──────────────────────────────────────────────────────────

    fun shutdown() {
        scope.cancel()
        httpClient.dispatcher.executorService.shutdown()
        httpClient.connectionPool.evictAll()
    }
}

// Suspend Cancellable
private suspend fun <T> Call.await(): Response = suspendCancellableCoroutine { cont ->
    cont.invokeOnCancellation { cancel() }
    enqueue(object : Callback {
        override fun onFailure(call: Call, e: IOException) { cont.resumeWithException(GenieError.NetworkError(e)) }
        override fun onResponse(call: Call, response: Response) { cont.resume(response) }
    })
}

// Shorthand
private inline fun <T> kotlinx.coroutines.CoroutineScope.e(crossinline block: suspend () -> T): T {
    return runBlocking { block() }
}

// Extension for executeRequest
private suspend inline fun <reified T> GenieClient.executeRequest(request: Request, clazz: Class<T>): T {
    val body = httpClient.newCall(request).await().body?.string() ?: ""
    return when {
        T::class == Unit::class -> Unit as T
        else -> json.decodeFromString(clazz.kotlin.javaObjectType, body)
    }
}

// GenieActivity stub (exported for manifest)
class GenieActivity : android.app.Activity()

// Re-export for convenience
typealias GenieClientInstance = GenieClient
